### ISSUE TITLE
New package to normalize spatial data for web plotting.

### ISSUE THREAD

**Issue by bhaskarvk** [r-spatial/discuss#15]
This is a bit of a future planning, but here is the main idea.
Currently there is code in the [leaflet](/rstudio/leaflet) package that extracts data from sp and sf objects and converts it into a dataframe that is then passed to the Javascript side (by converting it into a JSON).
This code is fairly generic and not really dependent on anything `leaflet` specific. It makes a lot of sense to take out this code and make it a package of its own. That way we can build other web plotting R packages to wrap say `d3.geo` or `mapboxGL` or `cesium` and reuse a major chunk of the code that takes  data from spatial objects and passes it to Javascript.

I have some discussions about this with @jcheng5 and agrees that this is a good idea. There are a coupe of questions I have for the r-spatial community.

a) Do you think this is a good idea ?
b) If so then do you think it makes sense for this proposed package to live in r-spatial repo ?
c) If b) is 'yes' what sort of licensing and copyright arrangement we need in place between RStudio and r-spatial ?

cc @tim-salabim @edzer

**Comment by tim-salabim** [comment]
@bhaskarvk 

a) I think this is a great idea
b) yes, I think this is very suited for r-spatial
c) I am no expert on licensing and all that is related, thus I don't really feel qualified to provide a definite answer on this one. I would love to get some more input [here](https://github.com/r-spatial/discuss/issues/2#issuecomment-323843777).

Two comments/questions from my side (that I can think of right now):
- I am not overly familiar with [`geojsonio`](https://github.com/ropensci/geojsonio) but isn't that what this package is about?
- Should we aim at implementing this in `C++` via `Rcpp` in order to handle large data performantly? Especially with regard to potentially implement webgl rendering.

**Comment by bhaskarvk** [comment]
@tim-salabim `geojsonio` is for reading/writing geo/topo JSONs from the file-system. What I am proposing is a common package that will take any spatial R object (sp, sf, geo/topo JSONs either as lists, char strings or R objects) and making them available to a htmlwidget in a consistent manner.
That way we can easily make new web GIS plotting pkgs that wrap mapboxGL / Cesium / OpenLayers etc.

It may well be that this package will rely on sp/sf/geojson/geojsonio packages to read the data but what differentiates it is the consistent manner in which it makes this spatial data available to the widget side.

So then you can have code that can look like 

```r

leaflet() %>% 
  addPolygons(some<sp|sf>polygon-data,...)

# OR

mapboxGL() %>%
  addPolygons(some<sp|sf>polygon-data,...)

# OR

openLayers() %>%
  addPolygons(some<sp|sf>polygon-data,...)

```

w/o having to duplicate the code which reads the spatial objects in individual leaflet/mapboxGL/Cesium packages.

**Comment by tim-salabim** [comment]
relevant https://github.com/rstudio/leaflet/issues/452

**Comment by edzer** [comment]
For the part (sf -> data.frame, sp -> data.frame), I think it makes more sense to have this as part of the sf and sp APIs, i.e. inside the packages.

For instance, https://github.com/rstudio/leaflet/issues/452 does not happen when leaflet uses `st_coordinates` on the `sfc` object, instead of calling `do.call(rbind, sfc)` which wrongly assumes `sfc` is not an empty list.

**Comment by bhaskarvk** [comment]
Yes that's an acceptable solution as well. I just think it belongs outside of leaflet.

**Comment by edzer** [comment]
Fair enough. Which functions in leaflet does this concern?

**Comment by mdsumner** [comment]
I've been working on this in spbabel and in a superseded form of that in https://github.com/mdsumner/sc  

The form must be relational, composed of multiple data frames - that's the only way to store all the types that are needed, and it's the only way to store topology at all (you can't do that with nesting, even if you nest indexes you still need a common pool table for those indexes to refer to). 

It's desperately needed that we have a common agreed form for these data in R, and I think specific packages should all contain decompositions to a generic form for their specific types. I've learnt enough in those and related projects for progressing my work but I'm very happy to pursue this in a more general form that transcends all and any specific implementations currently in use. The OSM work by rOpenSci has similar challenges, and osmdata in particular is an important use-case

@mpadge this is part of the general problem we've been talking about :)

**Comment by bhaskarvk** [comment]
@edzer All the code in leaflet's `R/normalize*.R` files is what I was thinking.

**Comment by SymbolixAU** [comment]
I might be going off on a tangent to the theme of this discussion, but, what's r-spatial/the community's thoughts on using [encoded polylines](https://developers.google.com/maps/documentation/utilities/polylinealgorithm) to represent geometries? 

Whenever I plot a map in `googleway` I always encode my spatial objects first as it reduces the size of the object being plotted (and the encoded polylines are natively supported in Google Map's API).

I've been playing about with a [spatialdatatable](https://github.com/SymbolixAU/spatialdatatable) package to do the encoding. I don't know how far I'm going to take this package, but if there's appetite to include it in r-spatial then I'll carry on.

**Comment by edzer** [comment]
I see this similar to [s2](https://github.com/r-spatial/sf/issues/502) cells and [geohash](https://en.wikipedia.org/wiki/Geohash); dedicated optimizations where you can afford some rounding and bandwidth is an issue. This one aims at communicating with the google maps stack. 

Your package does this, as well as the integration with data.table. Does it make sense to somehow separate that? 

If you believe it will attract a larger user community, we could move the package here.

**Comment by SymbolixAU** [comment]
I think separating the encoding/normalising is probably a good idea and would be a better fit for this 'new package' (whatever it turns out to be). And I think they'll definitely be a better way of doing the encoding from `sf` - to polyline that my nested `lapply`'s. I also started to look into `boost` and `CGAL`, but haven't progressed with it.
 

The reason I started writing `spatialdatatable` was to [speed up the `geosphere` calculations](https://stackoverflow.com/a/42014364/5977215), and also make them naturally usable inside `data.table[ ]` syntax.

**Comment by SymbolixAU** [comment]
I've made a start by creating [googlePolylines](https://github.com/SymbolixAU/googlePolylines) to handle the encoding and decoding of (primarily) `sf` objects into encoded polylines. As mentioned, the encoded lines reduce precision, but can [speed-up plotting](https://github.com/SymbolixAU/googlePolylines#motivation)  

I've seen plugins for leaflet to use these polylines too so there may be some opportunity for integration there too.

**Comment by SymbolixAU** [comment]
Thanks for the reminder @tim-salabim !

Given my recent updates to [mapdeck](https://github.com/SymbolixAU/mapdeck/issues/69) I think I've got a solid base of code to make this 'normalised data' package, so I'm happy to get this going.

anyone got a good suggestion for a package name?

**Comment by SymbolixAU** [comment]
[This is my idea of the package](https://github.com/SymbolixAU/spatialwidget)

**Comment by mdsumner** [comment]
@SymbolixAU can I suggest you take a look at [silicate](https://github.com/hypertidy/silicate/tree/binary) - the `binary` branch - there's two key functions `BINARY` and `SC`. 

* BINARY - builds a simple structure of `object` and `vertex` table, with edges of each object nested in `edge_`
* SC - builds a fully labelled-entity structure of `object`, `edge` and `vertex`

(`object` is `feature` in sf terms, but more general - we can have mesh types and other non SF forms)

The first is not topological (no vertex de-dupe) and cannot survive vertex subsetting without remapping the indexes. The second is topological (unique in x/y by default), with unique IDs for object, edge and vertex - so it can be arbitrarily ordered and passed through other systems. 

This has festered a bit, and my anglr package needs an update with the new SC/BINARY structure, but I'm hoping we can find common ground here. These forms admit conversion to other formats pretty easily, and there are verbs for extracting the entities `sc_coord`, `sc_path`, `sc_vertex` etc. so most of the format-specific details can go into methods for those.

**Comment by SymbolixAU** [comment]
yes. And I really want to start working with your structures to see what they are all about. We definitely need to get it all integrated.

**Comment by mdsumner** [comment]
all right, sorry for the dead horse flogging - spatialwidget doesn't look like what I thought you were talking about - trying to get a bearing on how you see things. :+1:

**Comment by SymbolixAU** [comment]
I'm going to add some concrete R examples to the spatialwidget package to hopefully make the design / rationale clear :)

**Comment by SymbolixAU** [comment]
I think under the hood it's along those lines, yes. With a little bit extra wrangling to create colours from the variables (and also a summary palette for a legend), and finally the geojson step. 

The idea is the output of `spatial_line()` feeds directly to javascript through the various `invoke_method()` calls in:
- [`addPolyline()` - leaflet](https://github.com/rstudio/leaflet/blob/master/R/layers.R#L1150)
- [`add_polyline()` - googleway](https://github.com/SymbolixAU/googleway/blob/master/R/google_map_layer_polyline.R#L195)
-  [`add_path()` - mapdeck](https://github.com/SymbolixAU/mapdeck/blob/master/R/map_layer_path.R#L94)

So internally, each of those `addPolyine()`, `add_polyline()`, `add_path()` will have a function body similar to

```r
add_new_polyline <- function(sf, ... ) {
  ## a bit of internal stuff for each implementation
  js <- spatial_line(...)
  invoke_method( ..., js , ... )
}
```

**Comment by mdsumner** [comment]
This is very nice helpful, can I ask about the dots passed on to invoke_method, is that some way of keeping track at R and js levels? (Or just some magic?)

**Comment by SymbolixAU** [comment]
Merely laziness on my part here, to indicate there are other arguments in those functions :)

**Comment by tim-salabim** [comment]
They are just additional arguments passed from R to the js method that is being invoked. 
See e.g. [here](https://github.com/tim-salabim/leaflet.glify/blob/master/R/glify-points.R#L91) for the R side and correspondingly [here](https://github.com/tim-salabim/leaflet.glify/blob/master/inst/htmlwidgets/Leaflet.glify/addGlifyPoints.js) for the js binding (the method) receives.

**Comment by mdsumner** [comment]
oh phew, thanks!

**Comment by SymbolixAU** [comment]
I've added three R functions, `widget_point()`, `widget_line()` and `widget_polygon()` which you can use directly, and I've updated the README and merged all the dev into master. 

I think this gives more concrete examples of what I'm aiming for. 

I'm going to use this `spatialwidget` library in mapdeck and googleway,  so that's been the primary focus of my design, but if there's anything leaflet/mapview would benefit from let me know.

**Comment by tim-salabim** [comment]
Cool! I will play with it in the near future. At the moment still focussing on leaflet.glify performance and usability enhancements.

**Comment by SymbolixAU** [comment]
To keep this thread updated, I'm planning on submitting `spatialwidget` to CRAN in a week or so

**Comment by harryprince** [comment]
I propose a detailed comparison between leafgl, deckgl and mapdeck to figure out which is the best solution when we need to plot large-scale points.

 https://github.com/r-spatial/leafgl/issues/11

**Comment by tim-salabim** [comment]
This discussion is partly related to #13. Hence, I'd be inclined to leave it open for now. 

I still haven't gotten around to play with `spatialwidget` but my feeling is that it is the closest we have come to a normalised spatial data package.
